Two small bugs in dual-stack init

thockin opened the pull request:
Working on the REST overhaul, adding testing turned these up:

1) Headless+Selectorless, on a single-stack cluster, policy=PreferDual

Prior to this commit, the result was a single IPFamiliy (because we
checked that the 2nd allocator was present).  This changes that case to
populate both families (we don't care if the allocator exists), which is
the same as RequireDual.

2) ClusterIP, user specifies 2 families but no IPs

Prior to this commit, the policy was inferred to be SingleStack.  This
changes that case to correctly default to RequireDual when 2 families
are present but no IPs.

```release-note
On single-stack configured (IPv4 or IPv6, but not both) clusters, Services which are both headless (no clusterIP) and selectorless (empty or undefined selector) will report `ipFamilyPolicy RequireDualStack` and will have entries in `ipFamilies[]` for both IPv4 and IPv6.  This is a change from alpha, but does not have any impact on the manually-specified Endpoints and EndpointSlices for the Service.
```

thockin committed:
Two small bugs in dual-stack init

Imporved testing turned these up:

1) Headless+Selectorless, on a single-stack cluster, policy=PreferDual

Prior to this commit, the result was a single IPFamiliy (because we
checked that the 2nd allocator was present).  This changes that case to
populate both families (we don't care if the allocator exists), which is
the same as RequireDual.

2) ClusterIP, user specifies 2 families but no IPs

Prior to this commit, the policy was inferred to be SingleStack.  This
changes that case to correctly default to RequireDual when 2 families
are present but no IPs.

aojea commented on pkg/registry/core/service/storage/rest_test.go:
```diff
@@ -3892,7 +3892,7 @@ func TestDefaultingValidation(t *testing.T) {
 			},
 			expectedIPFamilyPolicy: &preferDualStack,
-			expectedIPFamilies:     []api.IPFamily{api.IPv4Protocol},
+			expectedIPFamilies:     []api.IPFamily{api.IPv4Protocol, api.IPv6Protocol},
```
why if you "prefer" (not "require") and the cluster is single stack we have 2 ip families?

khenidak replied:
Headless, selector-less semantics.

thockin replied:
Basically, this is a corner case we can't comprehend from Service alone.  It could already be holding dual-stack endpoints (I know some users were doing this).  So we acknowledge that and admit that this Service could have both.  "require" vs "prefer" was what we specced - it felt more explicit.

aojea commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -902,15 +904,23 @@ func (rs *REST) tryDefaultValidateServiceClusterIPFields(service *api.Service) e
 
+	// Special-case: headless+selectorless
+	if len(service.Spec.ClusterIPs) > 0 && service.Spec.ClusterIPs[0] == api.ClusterIPNone && len(service.Spec.Selector) == 0 {
+		// If the use said nothing about policy and we can't infer it, they get dual-stack
```
this is not against the least surprise principle?
previous behaviour was the same as before dualstack

khenidak replied:
a case could be made that it has zero effect on user. As a matter of fact users already use that to create dual stack endpoints for non dualstack clusters today (before dual stack became a thing on kubernetes)

thockin replied:
Yeah, we don't enforce that they can't create Endpoints of any family.  At this case we just need to pick a value for the policy.  RequireDual was what we wrote in the KEP.

khenidak commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -872,18 +872,20 @@ func (rs *REST) tryDefaultValidateServiceClusterIPFields(service *api.Service) e
 
 		// we have previously validated for ip correctness and if family exist it will match ip family
 		// so the following is safe to do
+		//FIXME: once the 2nd-layer is removed, this could be a lie.
```
I am not sure i understand what is 2nd layer here?

aojea replied:
I think he refers to this https://github.com/kubernetes/kubernetes/pull/96684

thockin replied:
Sorry, forgot to remove this

khenidak commented:
nit: Please add release notes. Behavior is changing on single stack clusters.

re second part of the PR (`2 ClusterIPs` => `RequireDualStack`) -- we don't need to do that since it is already set in defaulting https://github.com/kubernetes/kubernetes/blob/e4e9c31218becac11f176cc824f5cc6b7a8036ac/pkg/apis/core/v1/defaults.go#L139-L142.

thockin reviewed (commented):
Stray comments removed.

thockin commented:
> we don't need to do that since it is already set in defaulting

That leaves a hole in the REST tests I am adding in the larger rework effort.  Maybe it makes sense to just do it all here and remove that defaulting?

thockin committed:
Remove defaults for ipFamilyPolicy and ipFamilies

These same values are set in the REST stack and it's easier to maintain
that code all in one place.

defaults.go is best for static values.

thockin commented:
I added a second commit which removes the code in defaults.go - it is all handled in REST anyway.

pacoxu commented:
/test pull-kubernetes-dependencies 
for go build github.com/aws/aws-sdk-go/service/ec2: /usr/local/go/pkg/tool/linux_amd64/compile: signal: killed (oom?)

aojea commented:
LGTM
defer to @khenidak

thockin commented:
/retest

khenidak commented:
/lgtm
/approve

thockin commented:
It looks like the go 1.15 is being used in the test?  When I run these locally they pass..

thockin commented:
Rebased

khenidak commented:
/lgtm
/approve
/retest

khenidak commented:
/test pull-kubernetes-e2e-kind-ipv6
